linked_list/linked_list.py

```diff
@@ -13,67 +12,157 @@ def __init__(self):
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
-
+        if self.head is None:
+            return None
+        else: 
+            return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
-
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
+        
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        current_node = self.head
+        while current_node is not None:
+            if current_node.value == value:
+                return True
+            current_node = current_node.next
+        return False
+            
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        node_count = 0
+        current_node = self.head
+        while current_node is not None:
+            node_count += 1
+            current_node = current_node.next
+        return node_count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        current_index = 0
+
+        if current_node is None:
+            return None
+        
+        while current_node is not None:
+            if current_index == index:
+                return current_node.value
+            else:
+                current_node = current_node.next
+                current_index+=1
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current_node = self.head
+
+        if current_node is None:
+            return None
+        
+        while True:

```

Comment:
Even if you know the loop will eventually exit, it's generally bad practice to use `while True`. With a while loop, we always want to make sure we are making progress inside the loop towards making the condition Falsy and that's impossible to do with `while True`

```suggestion
        while current_node is not None:

```